### PR TITLE
ASoC: SOF: add snd_sof_dsp_check_sdw_irq ops

### PR DESCRIPTION
**Pull request by bardliao**
sdw irq status checking is platform depence. Adding ops allow us to use
different sdw irq check function for each platform.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

### CODE REVIEW

**Review by ranj063 — DISMISSED** [review]

File: sound/soc/sof/intel/apl.c [hunk 1]
```diff
@@ -54,6 +54,9 @@ const struct snd_sof_dsp_ops sof_apl_ops = {
 	.ipc_msg_data	= hda_ipc_msg_data,
 	.ipc_pcm_params	= hda_ipc_pcm_params,
 
+	/* sdw */
+	.check_sdw_irq	= hda_dsp_check_sdw_irq,
```
Comment (plbossart) [inline sound/soc/sof/intel/apl.c]: SoundWire is not supported on APL, let's remove this.

File: sound/soc/sof/intel/apl.c [hunk 2]
```diff
@@ -54,6 +54,9 @@ const struct snd_sof_dsp_ops sof_apl_ops = {
 	.ipc_msg_data	= hda_ipc_msg_data,
 	.ipc_pcm_params	= hda_ipc_pcm_params,
 
+	/* sdw */
```
Comment (plbossart) [inline sound/soc/sof/intel/apl.c]: commit message:

SoundWire IRQ status checks are platform-dependent, add new ops structure to provide abstraction.

File: sound/soc/sof/intel/hda.c [hunk 3]
```diff
@@ -836,7 +836,7 @@ static irqreturn_t hda_dsp_interrupt_thread(int irq, void *context)
 	if (hda_dsp_check_ipc_irq(sdev))
 		sof_ops(sdev)->irq_thread(irq, sdev);
 
-	if (hda_dsp_check_sdw_irq(sdev))
+	if (snd_sof_dsp_check_sdw_irq(sdev))
```
Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: I have mixed feelings on this, is this really a generic function that others might use.
Wondering if this check_sdw_irq really belongs here.
Not opposed, just not sure if this is the right place to declare such a callback.

**Review by plbossart — CHANGES_REQUESTED** [review]
One nit-pick on whether this is really a generic platform capability, but I don't have an alternate solution to suggest.

File: sound/soc/sof/intel/hda.c (hunk 3 above)
Comment (bardliao) [reply sound/soc/sof/intel/hda.c]: We may move snd_sof_dsp_check_sdw_irq(sdev) to a better place once we find a better place for it.

File: sound/soc/sof/intel/apl.c (hunk 1 above)
Comment (bardliao) [reply sound/soc/sof/intel/apl.c]: updated

File: sound/soc/sof/intel/hda.c (hunk 3 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda.c]: Does hda_sdw_check_wakeen_irq need another ops for new platform ?  And can we merge ipc, sdw, wakeen irq into one ops then we don't need to have 3 ops for different irq.

Comment (bardliao) [reply sound/soc/sof/intel/hda.c]: > Does hda_sdw_check_wakeen_irq need another ops for new platform ? 

@RanderWang Looking at the code [#define HDA_DSP_REG_SNDW_WAKE_STS      0x2C192](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/hda.h#L236). I think it is actually [SDW_SHIM_BASE	(0x2C000)](https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/intel_init.c#L22) + [SDW_SHIM_WAKESTS	(0x192)](https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/intel.c#L64). We will have different sdw shim base, so some change is required. However, maybe we don't need a new ops.

>  And can we merge ipc, sdw, wakeen irq into one ops then we don't need to have 3 ops for different irq.

I would prefer to have separate ops since we have different handler. Or maybe a ops which return a flag to indicate which irq is triggered. Like 
```
#define STREAM_IRQ	BIT(0)
#define IPC_IRQ		BIT(1)
...
```
What do you think? @plbossart

Comment (plbossart) [reply sound/soc/sof/intel/hda.c]: Sorry I don't understand @RanderWang's comment nor @bardliao 's answer.
We have different sources of interrupts, and we need to check for each source and deal with it as appropriate.
Maybe we can handle Cadence and wakeen interrupts together, since they are both related to SoundWire.
But I don't see the rationale for handling streams or IPCs in the same ops as SoundWire, there are multiple cases where SoundWire is disabled. We should not mix things that are not connected?

Comment (bardliao) [reply sound/soc/sof/intel/hda.c]: @plbossart The new comments may solve your confusion. :)

File: sound/soc/sof/sof-priv.h [hunk 4]
```diff
@@ -162,6 +162,9 @@ struct snd_sof_dsp_ops {
 	int (*send_msg)(struct snd_sof_dev *sof_dev,
 			struct snd_sof_ipc_msg *msg); /* mandatory */
 
+	/* sdw */
+	bool (*check_sdw_irq)(struct snd_sof_dev *sdev); /* optional */
```
Comment (RanderWang) [inline sound/soc/sof/sof-priv.h]: I still feel that checking sdw irq is not very common in sof ops

Comment (bardliao) [reply sound/soc/sof/sof-priv.h]: > I still feel that checking sdw irq is not very common in sof ops

How about adding check_sdw_irq ops into [struct sof_intel_dsp_desc](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/shim.h#L155)? I will follow the majority. @plbossart @RanderWang @ranj063

Comment (plbossart) [reply sound/soc/sof/sof-priv.h]: using struct sof_intel_dsp_desc sounds good to me, you already have SSP information in there.

Comment (RanderWang) [reply sound/soc/sof/sof-priv.h]: sure, a good idea

File: sound/soc/sof/intel/cnl.c [hunk 5]
```diff
@@ -349,6 +349,8 @@ const struct sof_intel_dsp_desc cnl_chip_info = {
 	.rom_init_timeout	= 300,
 	.ssp_count = CNL_SSP_COUNT,
 	.ssp_base_offset = CNL_SSP_BASE_OFFSET,
+	.sdw_shim_base = SDW_SHIM_BASE,
```
Comment (plbossart) [inline sound/soc/sof/intel/cnl.c:352]: typo in commit message: different

File: sound/soc/sof/intel/hda.c [hunk 6]
```diff
@@ -129,6 +129,17 @@ int hda_ctrl_dai_widget_free(struct snd_soc_dapm_widget *w)
...
+	*get_chip_info(struct snd_sof_pdata *pdata)
+{
+	const struct sof_dev_desc *desc = pdata->desc;
+	const struct sof_intel_dsp_desc *chip_info;
+
+	chip_info = desc->chip_info;
+
+	return chip_info;
+}
+
```
Comment (plbossart) [inline sound/soc/sof/intel/hda.c:142]: maybe do this in a separate patch to prepare the change, then add the rest. It'll be easier to review.

**Review by plbossart — CHANGES_REQUESTED** [review]
looks good @bardliao minor nit-picks below

File: sound/soc/sof/intel/hda.c (hunk 6 above)
Comment (bardliao) [reply sound/soc/sof/intel/hda.c:142]: Done

**Review by RanderWang — APPROVED** [review]
LGTM

**Comment by bardliao** [comment]
@plbossart @RanderWang Added a new commit to use the shim/alh base on Intel SoundWire driver.

**Review by plbossart — APPROVED** [review]
thanks @bardliao

**Review by ranj063 — APPROVED** [review]

**Comment by marc-hb** [comment]
This does not compile for me, fixup in #3036
